pkg/px-resource-gateway/semaphore.go

```diff
@@ -0,0 +1,310 @@
+package server
```

Comment:
Please explain the algorithm as comment header in this file to help maintainers (and me as the reviewer)

Reply:
I'm skipping this file until the algorithm is included.